File: src/main/kotlin/Day02.kt
import java.io.File

val guide = File("src/main/resources/day02.txt")
    .readLines()
    .map { line ->
        val (a, b) = line.trim().split(" ")
        a.first() to b.first()
    }

sealed class RPSMove(val value: Int, val weakTo: RPSMove?, val strongerThan: RPSMove?) {
    abstract fun play(opponent: RPSMove): Int
}

class Rock: RPSMove(1, Paper(), Scissors())  {
    override fun play(opponent: RPSMove): Int {
        return value + when (opponent) {
            is Rock -> 3
            is Paper -> 0
            is Scissors -> 6
        }
    }
}

class Paper: RPSMove(2, Scissors(), Rock()) {
     override fun play(opponent: RPSMove): Int {
        return value + when (opponent) {
            is Rock -> 6
            is Paper -> 3
            is Scissors -> 0
        }
    }
}

class Scissors: RPSMove(3, Rock(), Paper()) {
    override fun play(opponent: RPSMove): Int {
        return value + when (opponent) {
            is Rock -> 0
            is Paper -> 6
            is Scissors -> 3
        }
    }
}

fun opponentMove(symbol: Char): RPSMove {
     return when(symbol) {
         'A' -> Rock()
         'B' -> Paper()
         'C' -> Scissors()
         else -> error("")
     }
}

fun firstReading(symbol: Char): RPSMove {
    return when(symbol) {
        'X' -> Rock()
        'Y' -> Paper()
        'Z' -> Scissors()
        else -> error("")
    }
}

fun secondReading(symbol: Char, opponent: RPSMove): RPSMove {
    return when(symbol) {
        'X' -> opponent.strongerThan ?: opponent
        'Y' -> opponent
        'Z' -> opponent.weakTo ?: opponent
        else -> error("")
    }
}

fun solvePart1() {
    val score = guide.sumOf { (their, mine) ->
        val opponentMove = opponentMove(their)
        val move = firstReading(mine)
        move.play(opponentMove)
    }
    println("Part 1. $score")
}

fun solvePart2() {
    var score = 0
    for ((their, mine) in guide) {
        val opponentMove = opponentMove(their)
        val move = secondReading(mine, opponentMove)
        score += move.play(opponentMove)
    }
    println("Part 2. $score")
}
fun main() {
    solvePart1()
    solvePart2()
}

File: 2022/Python/day02/day02.py
with open('day02.txt', 'r') as f :
    guide = []
    for line in f.readlines() :
        a, b = line.strip().split()
        opponentMove = a.replace('A', 'R').replace('B', 'P').replace('C', 'S')
        guide.append( (opponentMove, b) )


def singleTurnScore(move, opposingMove) :
    # Rock     -> 1
    # Paper    -> 2
    # Scissors -> 3
    myMoveScore = {'R': 1, 'P': 2, 'S': 3}

    myMove = myMoveScore[move]
    theirMove = myMoveScore[opposingMove]

    delta = myMove - theirMove

    if delta == 0 :
        return myMove + 3

    elif delta == 1 :
        return myMove + 6
    elif delta == -1 :
        return myMove
    elif delta == 2 :
        return myMove
    else :
        return myMove + 6


def firstReading(rule) :
    translation = {'X': 'R', 'Y': 'P', 'Z': 'S'}
    return translation[rule]


def secondReading(rule, opponent) :
    moveThatLosesTo = {'R': 'S', 'P': 'R', 'S': 'P'}
    moveThatWinsOver = {'R': 'P', 'P': 'S', 'S': 'R'}

    if   rule == 'X' :
        return moveThatLosesTo[opponent]
    elif rule == 'Z' :
        return moveThatWinsOver[opponent]
    else :
        return opponent



def solvePartOne() :
    score = 0
    for opponentMove, rule in guide :
        move = firstReading(rule)
        score += singleTurnScore(move, opponentMove)
    print(f"Part 1. {score}")


def solvePartTwo() :
    score = 0
    for opponentMove, rule in guide :
        move = secondReading(rule, opponentMove)
        score += singleTurnScore(move, opponentMove)
    print(f"Part 2. {score}")


solvePartOne()
solvePartTwo()